src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs

```diff
@@ -733,82 +643,20 @@ public Task SendMailAsync(MailMessage message, CancellationToken cancellationTok
                 return Task.FromCanceled(cancellationToken);
             }
 
-            // Create a TaskCompletionSource to represent the operation
-            var tcs = new TaskCompletionSource();
-
-            CancellationTokenRegistration ctr = default;
-
-            // Indicates whether the CTR has been set - captured in handler
-            bool ctrSet = false;
-
-            // Register a handler that will transfer completion results to the TCS Task
-            SendCompletedEventHandler? handler = null;
-            handler = (sender, e) =>
-            {
-                if (e.UserState == tcs)
-                {
-                    try
-                    {
-                        ((SmtpClient)sender).SendCompleted -= handler;
-                        if (Interlocked.Exchange(ref ctrSet, true))
-                        {
-                            // A CTR has been set, we have to wait until it completes before completing the task
-                            ctr.Dispose();
-                        }
-                    }
-                    catch (ObjectDisposedException) { } // SendAsyncCancel will throw if SmtpClient was disposed
-                    finally
-                    {
-                        if (e.Error != null) tcs.TrySetException(e.Error);
-                        else if (e.Cancelled) tcs.TrySetCanceled();
-                        else tcs.TrySetResult();
-                    }
-                }
-            };
-            SendCompleted += handler;
-
-            // Start the async operation.
-            try
-            {
-                SendAsync(message, tcs);
-            }
-            catch
-            {
-                SendCompleted -= handler;
-                throw;
-            }
-
-            ctr = cancellationToken.Register(s =>
-            {
-                ((SmtpClient)s!).SendAsyncCancel();
-            }, this);
+            Task t = SendAsyncInternal<AsyncReadWriteAdapter>(message, false, null, true, cancellationToken);
 
-            if (Interlocked.Exchange(ref ctrSet, true))
+            if (t.IsFaulted)
             {
-                // SendCompleted was already invoked, ensure the CTR completes before returning the task
-                ctr.Dispose();
+                // If the task completed unwrap the exception (if any)
+                t.GetAwaiter().GetResult();
```

Comment:
(Same for other places)

This is changing exception behavior since there's a race condition when checking `IsFaulted` - you don't know if it actually failed synchronously, or this thread just happened to observe an async exception by the time it got here.

You might need a `Task<Exception?>` or similar (e.g. splitting the async method into validation + real async).
If that's not feasible / unnecessarily complex, it'd IMO be better to wrap all exceptions than to potentially synchronously throw for more cases.